actions/ulog_ingestion/src/ulog_ingestion/utils.py

```diff
@@ -0,0 +1,287 @@
+                    topic.add_message_path(
+                        request=topics.AddMessagePathRequest(
+                            message_path=array_name,
+                            data_type=array_field_type,  # TBD
+                            canonical_data_type=topics.CanonicalDataType.Array,
+                        )
+                    )
+
+                    print(
+                        f"Adding sub-field for array: {sub_array_name}, type: {field.type_str}, canonical: {TYPE_MAPPING_CANONICAL[field.type_str]}"
+                    )
+
+                    # Add another message path for the array elements
+                    sub_array_name = f"{array_name}.[*]"
+                    topic.add_message_path(
+                        request=topics.AddMessagePathRequest(
+                            message_path=sub_array_name,
+                            data_type=field.type_str,  # TBD
+                            canonical_data_type=TYPE_MAPPING_CANONICAL[field.type_str],
+                        )
+                    )
+
+                    array_list.append(array_name)
+            else:
+                print(
+                    f"Adding field: {field.field_name}, type: {field.type_str}, canonical: {TYPE_MAPPING_CANONICAL[field.type_str]}"
+                )
+                topic.add_message_path(
+                    request=topics.AddMessagePathRequest(
+                        message_path=field.field_name,
+                        data_type=field.type_str,
+                        canonical_data_type=TYPE_MAPPING_CANONICAL[field.type_str],
+                    )
+                )
+
+    return
+
+
+def create_per_topic_mcap_from_ulog(
+    output_path_per_topic_mcap: Any, d: str, schema_registry_dict: Dict[str, Any]
+) -> None:
+    """
+    Creates a per-topic MCAP file from a ULog object.
+
+    Args:
+    - output_path_per_topic_mcap: The path to the output MCAP file.
+    - d: The ULog object containing the data to be converted.
+    - schema_registry_dict: A dictionary containing the JSON schemas for each topic.
+
+    Returns:
+    - None
+    """
+    with open(output_path_per_topic_mcap, "wb") as stream:
+        writer = Writer(stream)
+        writer.start()
+
+        schema_id = writer.register_schema(
+            name=d.name,
+            encoding="jsonschema",
+            data=json.dumps(schema_registry_dict[d.name]).encode(),
+        )
+
+        channel_id = writer.register_channel(
+            schema_id=schema_id,
+            topic=d.name,
+            message_encoding="json",
+        )
+
+        for i in range(len(d.data["timestamp"])):
+            values = list()
+            for f in d.field_data:
+                values.append((f.field_name, f.type_str, d.data[f.field_name][i]))
+            json_msg_instance = parse_values_to_json(values)
+            writer.add_message(
+                channel_id=channel_id,
+                log_time=int(d.data["timestamp"][i] * 1000),
```

Comment:
What's this converting to? Milliseconds? Nanoseconds? We should probably make a decision about what time resolution we're expecting.

Reply:
Agreed, this is Nanoseconds. I changed it to Nanoseconds because .mcap expects nanoseconds: https://mcap.dev/spec#timestamp. 

Changed the variable name to timestamp_ns to make it clear